### PR TITLE
feat: DLQ unprocessable messages on ingest-events

### PR DESCRIPTION
**Pull request by lynnagara**
This PR depends on https://github.com/getsentry/sentry-kafka-schemas/pull/230

It attempts to be somewhat cautious about DLQing and not DLQ
anything that is even potentially retriable. This could be tweaked
later.

If a non-retriable exception is raised, the consumer tries to further determine whether the message
is actually bad (or it's some transient retriable error) by validating the message against
the schema. The message will be DLQed only if that also fails.

### CODE REVIEW

File: src/sentry/ingest/consumer/simple_event.py [hunk 1]
```diff
@@ -29,26 +30,43 @@ def process_simple_event_message(
...
-    raw_payload = raw_message.payload.value
-    metrics.distribution(
-        "ingest_consumer.payload_size",
-        len(raw_payload),
-        tags={"consumer": consumer_type},
-        unit="byte",
-    )
-    message: IngestMessage = msgpack.unpackb(raw_payload, use_list=False)
+    try:
+        raw_payload = raw_message.payload.value
```
Comment (lynnagara) [inline src/sentry/ingest/consumer/simple_event.py:34]: none of this code changed, it's just indented in try/except block now

File: src/sentry/ingest/consumer/simple_event.py [hunk 2]
```diff
@@ -29,26 +30,43 @@ def process_simple_event_message(
...
+            try:
+                with metrics.timer("ingest_consumer.fetch_project"):
+                    project = Project.objects.get_from_cache(id=project_id)
+            except Project.DoesNotExist:
+                logger.exception("Project for ingested event does not exist: %s", project_id)
+                return
+
+            return process_event(message, project, reprocess_only_stuck_events)
+        except Exception as exc:
+            raise Retriable(exc)
```
Comment (lynnagara) [inline src/sentry/ingest/consumer/simple_event.py:59]: just to be safe and avoid dlqing anything potentially retriable, we don't dlq any exception from the `process_event` function

File: src/sentry/ingest/consumer/simple_event.py [hunk 3]
```diff
@@ -29,26 +30,43 @@ def process_simple_event_message(
...
 
-    return process_event(message, project, reprocess_only_stuck_events)
+    except Exception as exc:
+        # Non retriable exceptions raise InvalidMessage, which Arroyo will DLQ.
```
Comment (untitaker) [inline src/sentry/ingest/consumer/simple_event.py:62]: not sure this makes a ton of sense. why put retrieable errors in the DLQ if nobody can retry those DLQ items?

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:62]: It's the non-retriable errors not retriable ones that go in the DLQ, so the consumer does not get stuck on those. The retriable errors are things that might be caused by network issues, temporary unavailability of database, etc and should not be DLQed.

Comment (untitaker) [reply src/sentry/ingest/consumer/simple_event.py:62]: that's what I mean. IMO retriable errors should go into the DLQ as well. the DLQ is meant to be replayed.

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:62]: I think in an ideal world we'd do that eventually but I'm not sure if we're there yet, and we haven't tried it anywhere. IMO inspection of events and replaying needs to be built in a really fast and easy manner before taking this step.

Today DLQing everything in the case of temporary blips might cause a longer and more manual recovery period and comes with it's own risks that I haven't fully thought through yet.

Let's do it as a follow up later.

Comment (mwarkentin) [reply src/sentry/ingest/consumer/simple_event.py:62]: Would INC-660 messages have been marked as `Retriable` or `InvalidMessage` in this case?

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:62]: since the message was completely empty, it would've failed on the first line and raised `InvalidMessage` and been DLQ.  But more subtlely wrong messages might still slip through the cracks in the future. Note: this is really just a starting point to wire up a DLQ on basic failures, but we will probably want to tweak exactly what gets DLQed later.

Comment (mwarkentin) [reply src/sentry/ingest/consumer/simple_event.py:62]: Sounds good. I think I agree with @untitaker that generally any message that is failing should go into the DLQ, or we risk blocking up the pipeline again because of some unforeseen message or parsing issue. Perhaps with some built in retry and then DLQ after 3 failures or something. 

Building the tools to be able to reprocess easily from the DLQ sounds important too. 😊

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:62]: Are you sure you want to do this in this PR though, before the other tooling is built? I'm happy to change it now but be aware that SRE will potentially have to do large amount of manual consuming from DLQ even in the case of a 30 second temporary unavailability or we risk dropping messages if we experience some very fleeting network or availability issue

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:62]: In the meantime, I'm going to work on defining a schema for this topic so we can use that to determine what is valid or not.

Comment (mwarkentin) [reply src/sentry/ingest/consumer/simple_event.py:62]: No, what you have here seems good to start with, just that we should continue thinking about how to turn it into a general purpose failure handling tool.

Comment (jjbayer) [reply src/sentry/ingest/consumer/simple_event.py:62]: During INC-660, we crashed inside `process_event`, here: https://github.com/getsentry/sentry/blob/270094f47c840b6f7c084f5e569d3d61d9fda0e2/src/sentry/ingest/consumer/processors.py#L129

See [this sentry error](https://sentry.my.sentry.io/organizations/sentry/issues/683565/events/b1682b6ab8af4abda4fc2a11083b9a2e/).

The kafka message itself was populated, but the event JSON inside was empty.

@lynnagara How often does the consumer currently crash because of temporary network outages? If it's rare enough (say, once per month), I would personally go ahead and DLQ all exceptions.

I do believe this PR is an improvement (any DLQ is better than none), just pointing out that it would not have helped with INC-660.

Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py]: Ok, I realised it's going to be a lot of work to figure out what's retriable and what isn't on a consumer by consumer basis. So I rewrote this to check against the schema whenever an exception is raised to decide whether to DLQ it or not. Since there wasn't a schema for ingest-events, I added one here. https://github.com/getsentry/sentry-kafka-schemas/pull/230

**Comment by lynnagara** [comment]
@untitaker @jjbayer Can I pick on you two to review this? Would be nice to have some eyes from both the ingest + unified consumer perspective?

**Review by jjbayer — COMMENTED** [review]
Looks good to me, but should we add tests for the new code paths? E.g. one with an invalid message and one that mocks a connection error in `process_event`.

**Review by untitaker — APPROVED** [review]

**Review by jjbayer — APPROVED** [review]

File: src/sentry/consumers/__init__.py [hunk 4]
```diff
@@ -259,6 +259,7 @@ def ingest_events_options() -> list[click.Option]:
         "static_args": {
             "consumer_type": "events",
         },
+        "dlq_topic": Topic.INGEST_EVENTS_DLQ,
```
Comment (fpacifici) [inline src/sentry/consumers/__init__.py:262]: Has the topic already been created in all environments ?

File: src/sentry/ingest/consumer/simple_event.py [hunk 5]
```diff
@@ -12,6 +15,13 @@
 logger = logging.getLogger(__name__)
 
 
+consumer_type_to_default_topic = {
+    "events": Topic.INGEST_EVENTS,
+    "transactions": Topic.INGEST_TRANSACTIONS,
+    "attachments": Topic.INGEST_ATTACHMENTS,
+}
+
+
```
Comment (fpacifici) [inline src/sentry/ingest/consumer/simple_event.py:24]: This does not cover the attachment topic. 
The functions for attachments are `process_attachments_and_events` and `decode_and_process_chunks`. 
I think we should cover all of them. Whether you want to do attachment in a separate PR that's ok.

File: src/sentry/ingest/consumer/simple_event.py [hunk 6]
```diff
@@ -36,19 +46,45 @@ def process_simple_event_message(
...
+        if default_topic != "ingest-events":
+            raise
+
+        codec = sentry_kafka_schemas.get_codec(default_topic)
+
+        try:
+            codec.decode(raw_payload, validate=True)
+        except Exception:
+            raw_value = raw_message.value
+            assert isinstance(raw_value, BrokerValue)
```
Comment (fpacifici) [inline src/sentry/ingest/consumer/simple_event.py:86]: Can it be anything else ?

File: src/sentry/ingest/consumer/simple_event.py [hunk 7]
```diff
@@ -36,19 +46,45 @@ def process_simple_event_message(
...
+
+        default_topic = consumer_type_to_default_topic[consumer_type].value
+
+        # TODO: Currently, there is only a schema for ingest-events, so just continue to re-raise
+        # the exception if it's a different topic. This can be removed once attachments and transactions
+        # have schemas too.
+        if default_topic != "ingest-events":
+            raise
+
+        codec = sentry_kafka_schemas.get_codec(default_topic)
```
Comment (fpacifici) [inline src/sentry/ingest/consumer/simple_event.py:80]: Isn't this a heavy operation ? Any reason not to instantiate the codec only once ?

File: src/sentry/ingest/consumer/simple_event.py [hunk 8]
```diff
@@ -36,19 +46,45 @@ def process_simple_event_message(
...
+        try:
+            codec.decode(raw_payload, validate=True)
+        except Exception:
+            raw_value = raw_message.value
+            assert isinstance(raw_value, BrokerValue)
+
+            raise InvalidMessage(raw_value.partition, raw_value.offset)
 
-    return process_event(message, project, reprocess_only_stuck_events)
+        raise
```
Comment (fpacifici) [inline src/sentry/ingest/consumer/simple_event.py:90]: I thought we were discussing about inverting the way DLQ classifies retriable errors: Instead of putting the burden on the consumer developer to identify errors that should make us route the message to the DLQ, asking the consumer developers to identify errors that should trigger a crash and treat everything else as InvalidMessage.

Is that still the case and are you planning to deal with this as a separate PR/project ?

File: src/sentry/ingest/consumer/simple_event.py (hunk 7 above)
Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:80]: it's cached in the library

File: src/sentry/ingest/consumer/simple_event.py (hunk 6 above)
Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:86]: no

File: src/sentry/ingest/consumer/simple_event.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:24]: Yes. One step at a time.

File: src/sentry/ingest/consumer/simple_event.py (hunk 8 above)
Comment (lynnagara) [reply src/sentry/ingest/consumer/simple_event.py:90]: This is still case, but in this specific function it's very difficult because of the specific logic / external systems that are depended and things that can fail here. Practically, I think checking for both an exception + then the schema as well is the safest way in this one case.

File: src/sentry/consumers/__init__.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/consumers/__init__.py:262]: yes

**Comment by lynnagara** [comment]
@jjbayer I brought back the retriable exception around the processing store blocks to handle the case where message still passes the schema but the inner part of the payload is invalid
